redis/src/tls.rs

```diff
@@ -0,0 +1,160 @@
+///         &mut reader_ca
+///     )
+///     .expect("Unable to build client");
+///
+///     let connection_info = client.get_connection_info();
+///
+///     println!(">>> connection info: {connection_info:?}");
+///
+///     let mut con = client.get_async_connection().await?;
+///
+///     con.set("key1", b"foo").await?;
+///
+///     redis::cmd("SET")
+///         .arg(&["key2", "bar"])
+///         .query_async(&mut con)
+///         .await?;
+///
+///     let result = redis::cmd("MGET")
+///         .arg(&["key1", "key2"])
+///         .query_async(&mut con)
+///         .await;
+///     assert_eq!(result, Ok(("foo".to_string(), b"bar".to_vec())));
+///     println!("Result from MGET: {result:?}");
+///
+///     Ok(())
+/// }
+/// ```
+pub fn build_with_tls<C: IntoConnectionInfo>(
+    conn_info: C,
+    client_cert: &mut dyn BufRead,
+    client_key: &mut dyn BufRead,
+    root_cert: &mut dyn BufRead,
+) -> RedisResult<Client> {
+    let mut connection_info = conn_info.into_connection_info()?;
+    let tls_params = retrieve_tls_certificates(client_cert, client_key, root_cert)?;
+
+    connection_info.addr = if let ConnectionAddr::TcpTls {
+        host,
+        port,
+        insecure,
+        ..
+    } = connection_info.addr
+    {
+        ConnectionAddr::TcpTls {
+            host,
+            port,
+            insecure,
+            tls_params: Some(tls_params),
+        }
+    } else {
+        return Err(RedisError::from((
+            ErrorKind::InvalidClientConfig,
+            "TLS client requires a `rediss://` address",
+        )));
+    };
+
+    Ok(Client { connection_info })
+}
+
+pub(crate) fn retrieve_tls_certificates(
+    mut client_cert: &mut dyn BufRead,
+    client_key: &mut dyn BufRead,
+    mut root_cert: &mut dyn BufRead,
+) -> RedisResult<TlsConnParams> {
+    let client_cert: Vec<Certificate> =
+        certs(&mut client_cert).map(|mut certs| certs.drain(..).map(Certificate).collect())?;
+
+    let client_key = load_key(client_key)?;
+
+    let certs = certs(&mut root_cert)?;
+
+    let trust_anchors: RedisResult<Vec<OwnedTrustAnchor>> = certs
+        .iter()
+        .map(|cert| {
+            let ta = webpki::TrustAnchor::try_from_cert_der(cert)
+                .map_err(|_| Error::new(IOErrorKind::Other, "Unable to parse CA certificate"))?;
+
+            Ok(OwnedTrustAnchor::from_subject_spki_name_constraints(
+                ta.subject,
+                ta.spki,
+                ta.name_constraints,
+            ))
+        })
+        .collect();
+    let trust_anchors = trust_anchors?;
+
+    let mut root_cert = RootCertStore::empty();
+    root_cert.add_trust_anchors(trust_anchors.into_iter());
+
+    Ok(TlsConnParams {
+        client_key,
+        client_cert,
+        root_cert,
+    })
+}
+
+#[derive(Debug, Clone)]
+pub struct TlsConnParams {
+    pub(crate) client_key: PrivateKey,
+    pub(crate) client_cert: Vec<Certificate>,
+    pub(crate) root_cert: RootCertStore,
```

Comment:
Can `root_cert` be optional? I.e., is it possible that the CA cert is installed at the OS level and that we therefore don't need to pass it here?

Reply:
I guess it's possible, but not the use case I've seen around. 
My idea is to tackle the case in which we create a client only if all certificates/key are consistent in a single method, instead of relying on a builder with multiple steps to take and check consistence of certificates and keys.

Reply:
This is part of the public API, though, so if it's ever feasible that a client may retrieve the root CA certs from the OS, then I think we should leave open that possibility. The only changes I see required for that would be modifying `create_rustls_config` to use `root_store` instead in `with_with_certificates` if `tls_params.root_cert` is `None`, e.g.,

```
    let config = if let Some(tls_params) = tls_params {
        config
            .with_root_certificates(tls_params.root_cert.unwrap_or(root_store))
           ...
    } else { 
    ...
```

Practically speaking this might not work at the moment *anway*, b/c I don't think `rustls` will check OS-level certs unless we make some additional changes. But we can test/fix that later if needed. Right now I just want to make sure we're defining the public API the best way possible.

Reply:
Actually I'll turn my point on its head as well. Is it not possible we might want to enable explicitly specifying a specific root certificate, but don't actually have client-side auth enabled (i.e., `tls-auth-clients` is `no`) so don't care about supplying client cert and key?

Reply:
this way I see the possible scenarios are:
- TLS using CA in trust store (already implemented)
- TLS using CA provided 
- mTLS using CA in trust store and client cert/key
- mTLS using provided CA cert + client cert/key (what we have at [this](https://github.com/redis-rs/redis-rs/pull/858/commits/8aa5943d21545a0946930bfbf21f1b86f8194204#diff-8ce03c631ce32ec6617737c96699023373f6d8d35700bf9820a8b32951cb8e54R60) point)

if that's the case, maybe it would make sense to use a Client ConfigBuilder approach similar to (rustls client builder here)[https://github.com/redis-rs/redis-rs/blob/8aa5943d21545a0946930bfbf21f1b86f8194204/redis/src/connection.rs#L790-L792]

Reply:
Well, why not just make the client artifacts their own type, which would enforce the conditions for each potential scenario? E.g.,
```
struct ClientTlsParams {
    client_key: PrivateKey,
    client_cert: Vec<Certificate>,
}

pub struct TlsConnParams {
    client_params: Option<ClientTlsParams>,
    root_cert: Option<RootCertStore>,
}
```

Reply:
sorry, took few days off and only got some time to work on this couple days ago.

I've changed both `Client::build_with_tls` and `ClusterClientBuilder` to make `root_cert` and `client_tls` pair optional at the API, which allow setting TLS parameters as you suggested. 

Pls let me know what you think :+1: